core/src/main/scala/chisel3/Module.scala

```diff
@@ -382,6 +384,17 @@ package experimental {
       names
     }
 
+    /** Invokes _onModuleClose on HasIds found via reflection but not bound to hardware
+      * (thus not part of _ids)
+      * This maintains old naming behavior for non-hardware Data
+      */
+    private[chisel3] def closeUnboundIds(names: HashMap[HasId, String]): Unit = {
+      val idLookup = _ids.toSet
+      for ((id, _) <- names if !idLookup(id)) {
+        id._onModuleClose
+      }
+    }
```

Comment:
It would seem like this is undoing the work of this PR by promoting `HasId`s which we don't care about to not be GC-able by unnecessarily linking them to their parent module. 

What's going on here? I expect I'm missing something...

Reply:
This doesn't actually undo the work of this PR for 2 reasons:

1. it doesn't actually make any `HasIds` not GC-able. [_onModuleClose](https://github.com/chipsalliance/chisel3/blob/2ed343e2305b7c22000f3f46fa81d73a369907eb/core/src/main/scala/chisel3/Aggregate.scala#L648) in Record sets the refs of the elements of an unbound Record. This just provides references from the elements of the `Record` to the Record itself, there's a circular dependency there but it doesn't involve the Module at all which is where the real non-GCable Data comes from.
2. Because these are the `names` that come from reflective naming, the Module has a reference to each of these `HasIds` *anyway*, thus they aren't GC-able.

A follow on question is why not just add them to `_ids` and let the `for (id <- getIds) {` logic above handle calling `_onModuleClose`? I thought about it, but my reasoning is that it's clearer Iif `_ids` only contains the `HasIds` we care about and instead this logic (which is maintaining and old, questionable API) is intentionally kept separate. That's my thinking at least.